### ISSUE TITLE
Unintended hyperparameter search space changes v1->v2 ?

### ISSUE THREAD

**Issue by KnathanM** [chemprop/chemprop#1207]
I was looking at some of the PRs where we changed the hyperparameter search space. 

We changed the lower bound for learning rate hparams in #873 
We changed the dropout rate chances for being chosen in #734, based on #496 and #571
We also changed the batch size options in #734 and #873

But when I plot the pdfs for the hyperparameters in v1/v2 I see some unexpected differences. These stem from `tune.quniform(q=1)` not giving rounded floats so we used `tune.randint`, but that doesn't include the upper bound. Also note the somewhat surprising result that `quniform` doesn't give uniform probability to each option,. For each hparam, I sampled `n` samples from the given distribution in v1 (Hyperopt) and in v2 (Ray Tune) and plotted the choices vs how many times they were chosen. 

<details>
  <summary>Activation - No Change</summary>

`n=1200`
Note we are considering removing `SELU` from the search space in #1146 

![Image](https://github.com/user-attachments/assets/feff4316-f3bf-42a9-a1b6-da50f00b5560)

</details>

<details>
  <summary>Aggregation - No Change</summary>

`n=600`
We could add our attention aggregation to the search space by adding it to the aggregation registry.

![Image](https://github.com/user-attachments/assets/fd4ced2c-0fda-4277-90c5-e0351114d8d7)

</details>

<details>
  <summary>Aggregation Norm - Changed</summary>

`n=60000`
V2 uses `tune.quniform(lower=1, upper=200, q=1)`, which you would think is the same as V1's `hp.quniform("aggregation_norm", low=1, high=200, q=1)`, but in ray tune, if you use a quantized sampler with `q=1`, then the quantized part is ignored. (It checks `if self.q == 1` [here](https://github.com/ray-project/ray/blob/a9f8dc8c199a4ec73300816be79329f42673364a/python/ray/tune/search/sample.py#L552-L553).) So our current implementation will return a float like 57.345635664 for example. This seems fine to me for the aggregation norm. But if we want to keep the v1 behavior we could use `tune.qrandint(lower=1, upper=201, q=1)` instead. (This is equivalent to `tune.randint(lower=1, upper=201)`.) Note that Ray Tune's orange dots form a line at y=1 from 1 to 200 because each sampled number is not rounded so it is very unlikely any number will be sampled twice. 

![Image](https://github.com/user-attachments/assets/c9447928-0c22-4c12-99bd-de5f2b6ecfdf)

Note that Hyperopt gives less probability to the max and min. I'll discuss that under depth.

</details>

<details>
  <summary>Batch Size - Intentional Changed</summary>

`n=10000`
We changed the batch size to choose from powers of two because that felt more "computery". 

![Image](https://github.com/user-attachments/assets/afac1c02-0d9b-47ff-b958-fdb59f7e9bab)

</details>

<details>
  <summary>Depth - Changed</summary>

`n=10000`
V1 used `hp.quniform("depth", low=2, high=6, q=1)`, while V2 uses `tune.qrandint(lower=2, upper=6, q=1)`. (Note the change from `hp.quiniform(q=1)` to `tune.qrandint(q=1)` to keep integer outputs. We could have done `tune.randint` instead of `tune.qrandint(q=1)`.) But `tune.qrandint` doesn't include the upper bound, so the max for v2 is `depth=5` instead of `depth=6`. 

Using `tune.randint` (or the equivalent `tune.qrandint(q=1)`) pulls options with equal chance, so in v2 `depth=2` is chosen just as often as `depth=3`. But `hp.quniform` (and `tune.quniform(q!=1)`) don't do that. This is because the value is first sampled from a uniform distribution between low and high and then transformed into a quantized value using 

$$ y = round\left(\frac{x}{q}\right) * q $$

Say we did `hp.quniform(lower=0, upper=2, q=1)`. All sampled values between 0.5 and 1.5 will get mapped to 1, while values between 0 and 0.5 will get mapped to 0, leading to 0 being chosen less often. Based on conversations with others, it seems like we expected `quniform` to be uniform (e.g. in #571 it is claimed that the probability of droput being 0 in v1 was 1/9, when in fact it is 1/16, see dropout). 

I think we should change the depth search space back to include 6 as a possible value. I am not sure if we should keep the probability for each depth equal as it is now in v2, or change it back to the v1 behavior of the max and min having a lower probability of being chosen. The v1 behavior is not what we expect, but it is what has always been used. 

![Image](https://github.com/user-attachments/assets/7751fa27-28a3-495c-a09a-254652bfade3)

</details>

<details>
  <summary>Dropout - Intentional Changed</summary>

`n=3200`
The v2 probability is what we want. The v1 probability shows the same decreased odds for the max and min as was discussed under depth. 

![Image](https://github.com/user-attachments/assets/19f10d5a-e032-4e9a-9d1b-4358806c367f)

</details>

<details>
  <summary>FFN Hidden Dim - Maybe Needs Change</summary>

`n=10000`
Should we change this to have 300 and 2400 have equal probability? In this case, we are using `tune.quniform` with `q` not equal to 1 so it actually quantizes the uniform distribution the same as Hyperopt. 

![Image](https://github.com/user-attachments/assets/8eb27b20-905e-4aaf-9972-9ca0e303ba23)

</details>

<details>
  <summary>FFN Num Layers - Changed</summary>

`n=10000`
Same problems as depth.  Also see this issue #1183, that describes how FFN Num Layers in v2 means one more layer than v1. We have found that the FFN should have at least 1 hidden layer, which in v1 was given by `n_layers=2` while in v2 it is given by `n_layers=1`. So ultimately the change here is that in v2 hpopt won't consider a model with 0 hidden layers while in v1 it would. That seems like a good change to me. 

![Image](https://github.com/user-attachments/assets/78a338a0-014c-4178-8eb9-6cc58127a3cf)

</details>

<details>
  <summary>Final LR Ratio - Intentionally Changed</summary>

`n=10000`

![Image](https://github.com/user-attachments/assets/2611bc79-c762-47d1-86e5-f539f0ad827e)

</details>

<details>
  <summary>Message Hidden Dim - Maybe Needs Change</summary>

`n=10000`
Same as FFN Hidden Dim

![Image](https://github.com/user-attachments/assets/35d79e11-a7da-4fe0-895d-aee690d231a6)

</details>

<details>
  <summary>Init LR Ratio - Intentionally Changed</summary>

`n=10000`

![Image](https://github.com/user-attachments/assets/9c9a3f9c-01c2-4769-9c8c-8438b095fd00)

</details>

<details>
  <summary>Max LR Ratio - Intentionally Changed</summary>

`n=10000`

![Image](https://github.com/user-attachments/assets/867d7906-1050-473c-a560-61ae34c5cea6)

</details>

<details>
  <summary>Warmup Epochs - Changed</summary>

`n=10000`
Same as depth

![Image](https://github.com/user-attachments/assets/f3bdd69d-f422-48fb-b63a-02069316c75e)

</details>

<details>
  <summary>Code for plots</summary>

  ```python
import numpy as np
import matplotlib.pyplot as plt
from collections import Counter
from hyperopt import hp
from hyperopt.pyll.stochastic import sample
from ray import tune
from chemprop.nn import Activation, AggregationRegistry

def compare_distributions(name, hp_sampler, tune_sampler):
    try:
        n_options = len(tune_sampler)
    except TypeError:
        n_options = 50
    if name == "aggregation_norm":
        n = 60000
    elif name == "batch_size":
        n = 10000
    else:
        n = 200 * n_options
    hp_samples = [sample(hp_sampler) for _ in range(n)]
    tune_samples = [tune_sampler.sample() for _ in range(n)]
    print(n)
    if name == "activation":
        hp_samples = [s.lower() for s in hp_samples]
        tune_samples = [s.lower() for s in tune_samples]

    hp_counts = Counter(hp_samples)
    tune_counts = Counter(tune_samples)

    plt.figure(figsize=(8, 4))
    if isinstance(hp_samples[0], str):
        labels = sorted(set(hp_samples + tune_samples))
        hp_values = np.arange(len(labels))
        tune_values = hp_values
        hp_freq = [hp_counts.get(label, 0) for label in labels]
        tune_freq = [tune_counts.get(label, 0) for label in labels]
        plt.xticks(hp_values, labels)
    
    else:
        hp_values = sorted(hp_counts)
        hp_freq = [hp_counts[v] for v in hp_values]
        tune_values = sorted(tune_counts)
        tune_freq = [tune_counts[v] for v in tune_values]
        min_val = min(hp_samples)
        max_val = max(hp_samples)

        plt.text(0.5, 0.6, "HP Max: " + str(max_val) + "\nHP Min: " + str(min_val), transform=plt.gca().transAxes,
         verticalalignment='top', fontsize=12, bbox=dict(facecolor='white', alpha=0.7, edgecolor='gray'))
        tune_min_val = min(tune_samples)
        tune_max_val = max(tune_samples)
        plt.text(0.5, 0.4, "Tune Max: " + str(tune_max_val) + "\nTune Min: " + str(tune_min_val), transform=plt.gca().transAxes,
         verticalalignment='top', fontsize=12, bbox=dict(facecolor='white', alpha=0.7, edgecolor='gray'))


    if "lr" in name:
        plt.xscale("log")
    plt.scatter(hp_values, hp_freq, s=80, label="Hyperopt", color="blue")
    plt.scatter(tune_values, tune_freq, s=80, label="Ray Tune", color="orange")
    plt.legend()
    plt.title(name)
    plt.ylim(bottom=0)
    plt.show()

v2 = {
        "activation": tune.choice(categories=list(Activation.keys())),
        "aggregation": tune.choice(categories=list(AggregationRegistry.keys())),
        "aggregation_norm": tune.quniform(lower=1, upper=200, q=1),
        "batch_size": tune.choice([16, 32, 64, 128, 256]),
        "depth": tune.qrandint(lower=2, upper=6, q=1),
        "dropout": tune.choice([0.0] * 8 + list(np.arange(0.05, 0.45, 0.05))),
        "ffn_hidden_dim": tune.qrandint(lower=300, upper=2400, q=100),
        "ffn_num_layers": tune.qrandint(lower=1, upper=3, q=1),
        "final_lr_ratio": tune.loguniform(lower=1e-2, upper=1),
        "message_hidden_dim": tune.qrandint(lower=300, upper=2400, q=100),
        "init_lr_ratio": tune.loguniform(lower=1e-2, upper=1),
        "max_lr": tune.loguniform(lower=1e-4, upper=1e-2),
        "warmup_epochs": tune.qrandint(lower=1, upper=50 // 2, q=1),
    }

v1 = {
        "activation": hp.choice(
            "activation", options=["ReLU", "LeakyReLU", "PReLU", "tanh", "SELU", "ELU"]
        ),
        "aggregation": hp.choice("aggregation", options=["mean", "sum", "norm"]),
        "aggregation_norm": hp.quniform("aggregation_norm", low=1, high=200, q=1),
        "batch_size": hp.quniform("batch_size", low=5, high=200, q=5),
        "depth": hp.quniform("depth", low=2, high=6, q=1),
        "dropout": hp.quniform("dropout", low=0.0, high=0.4, q=0.05),
        "ffn_hidden_size": hp.quniform("ffn_hidden_size", low=300, high=2400, q=100),
        "ffn_num_layers": hp.quniform("ffn_num_layers", low=1, high=3, q=1),
        "final_lr_ratio": hp.loguniform("final_lr_ratio", low=np.log(1e-4), high=0.),
        "hidden_size": hp.quniform("hidden_size", low=300, high=2400, q=100),
        "init_lr_ratio": hp.loguniform("init_lr_ratio", low=np.log(1e-4), high=0.),
        "linked_hidden_size": hp.quniform("linked_hidden_size", low=300, high=2400, q=100),
        "max_lr": hp.loguniform("max_lr", low=np.log(1e-6), high=np.log(1e-2)),
        "warmup_epochs": hp.quniform("warmup_epochs", low=1, high=50 // 2, q=1)
    }

renames = {"ffn_hidden_dim": "ffn_hidden_size", "message_hidden_dim": "hidden_size"}

k = "activation"
compare_distributions(k, v1[renames.get(k, k)], v2[k])

k = "aggregation"
compare_distributions(k, v1[renames.get(k, k)], v2[k])

k = "aggregation_norm"
compare_distributions(k, v1[renames.get(k, k)], v2[k])

k = "batch_size"
compare_distributions(k, v1[renames.get(k, k)], v2[k])

k = "depth"
compare_distributions(k, v1[renames.get(k, k)], v2[k])

k = "dropout"
compare_distributions(k, v1[renames.get(k, k)], v2[k])

k = "ffn_hidden_dim"
compare_distributions(k, v1[renames.get(k, k)], v2[k])

k = "ffn_num_layers"
compare_distributions(k, v1[renames.get(k, k)], v2[k])

k = "final_lr_ratio"
compare_distributions(k, v1[renames.get(k, k)], v2[k])

k = "message_hidden_dim"
compare_distributions(k, v1[renames.get(k, k)], v2[k])

k = "init_lr_ratio"
compare_distributions(k, v1[renames.get(k, k)], v2[k])

k = "max_lr"
compare_distributions(k, v1[renames.get(k, k)], v2[k])

k = "warmup_epochs"
compare_distributions(k, v1[renames.get(k, k)], v2[k])
  ```
</details>

**Comment by KnathanM** [comment]
Thank you for the thoughts. 

[tune.randint](https://docs.ray.io/en/latest/tune/api/doc/ray.tune.randint.html) does not include the upper bound. The reason `tune.qrandint` does include the upper bound is because numbers close to the upper bound will get rounded to the upper bound, but the underlying randint sampler will not chose the upper bound identically. Because `qrandint` does not actually round if `q=1`, `qrandint(q=1)` behaves the same as `randint`. 

I'm a bit unsure what to do for FFN Num Layers. I'm leaning towards not changing it.

See also the discussion starting with this comment and down: https://github.com/chemprop/chemprop/pull/796#issuecomment-2059600183

**Comment by KnathanM** [comment]
Closed by #1230
